### PR TITLE
Fix(text2image): Return providers as indexed array

### LINKED ISSUE #21: Undefined array key 0 at /var/www/nextcloud/lib/private/TextToImage/Manager.php#231
NC 28.0.1

Installed:
  - text2image_helper: 1.0.1
  - text2image_stablediffusion: 1.0.1

In the admin settings everything is green. At first I thought its because the downloaded model had only 400MB, but after re-downloading the model, the same error still exists.

```
{
    "reqId": "nHI3pbXAIOyRY8OJD4iL",
    "level": 3,
    "time": "2023-12-18T21:49:02+01:00",
    "remoteAddr": "172.31.2.89",
    "user": "marcel.mueller",
    "app": "PHP",
    "method": "POST",
    "url": "/apps/text2image_helper/process_prompt",
    "message": "Undefined array key 0 at /var/www/nextcloud/lib/private/TextToImage/Manager.php#231",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "version": "28.0.1.0",
    "data": {
        "app": "PHP"
    }
}
```

```
{
    "reqId": "nHI3pbXAIOyRY8OJD4iL",
    "level": 3,
    "time": "2023-12-18T21:49:02+01:00",
    "remoteAddr": "172.31.2.89",
    "user": "marcel.mueller",
    "app": "index",
    "method": "POST",
    "url": "/apps/text2image_helper/process_prompt",
    "message": "{\"Exception\":\"Exception\",\"Message\":\"Call to a member function getExpectedRuntime() on null in file '/var/www/nextcloud/lib/private/TextToImage/Manager.php' line 231\",\"Code\":0,\"Trace\":[{\"file\":\"/var/www/nextcloud/lib/private/AppFramework/App.php\",\"line\":184,\"function\":\"dispatch\",\"class\":\"OC\\\\AppFramework\\\\Http\\\\Dispatcher\",\"type\":\"->\"},{\"file\":\"/var/www/nextcloud/lib/private/Route/Router.php\",\"line\":315,\"function\":\"main\",\"class\":\"OC\\\\AppFramework\\\\App\",\"type\":\"::\"},{\"file\":\"/var/www/nextcloud/lib/base.php\",\"line\":1069,\"function\":\"match\",\"class\":\"OC\\\\Route\\\\Router\",\"type\":\"->\"},{\"file\":\"/var/www/nextcloud/index.php\",\"line\":39,\"function\":\"handleRequest\",\"class\":\"OC\",\"type\":\"::\"}],\"File\":\"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php\",\"Line\":169,\"Previous\":{\"Exception\":\"Error\",\"Message\":\"Call to a member function getExpectedRuntime() on null\",\"Code\":0,\"Trace\":[{\"file\":\"/var/www/nextcloud/apps/text2image_helper/lib/Service/Text2ImageHelperService.php\",\"line\":91,\"function\":\"runOrScheduleTask\",\"class\":\"OC\\\\TextToImage\\\\Manager\",\"type\":\"->\"},{\"file\":\"/var/www/nextcloud/apps/text2image_helper/lib/Controller/Text2ImageHelperController.php\",\"line\":44,\"function\":\"processPrompt\",\"class\":\"OCA\\\\Text2ImageHelper\\\\Service\\\\Text2ImageHelperService\",\"type\":\"->\"},{\"file\":\"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php\",\"line\":230,\"function\":\"processPrompt\",\"class\":\"OCA\\\\Text2ImageHelper\\\\Controller\\\\Text2ImageHelperController\",\"type\":\"->\"},{\"file\":\"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php\",\"line\":137,\"function\":\"executeController\",\"class\":\"OC\\\\AppFramework\\\\Http\\\\Dispatcher\",\"type\":\"->\"},{\"file\":\"/var/www/nextcloud/lib/private/AppFramework/App.php\",\"line\":184,\"function\":\"dispatch\",\"class\":\"OC\\\\AppFramework\\\\Http\\\\Dispatcher\",\"type\":\"->\"},{\"file\":\"/var/www/nextcloud/lib/private/Route/Router.php\",\"line\":315,\"function\":\"main\",\"class\":\"OC\\\\AppFramework\\\\App\",\"type\":\"::\"},{\"file\":\"/var/www/nextcloud/lib/base.php\",\"line\":1069,\"function\":\"match\",\"class\":\"OC\\\\Route\\\\Router\",\"type\":\"->\"},{\"file\":\"/var/www/nextcloud/index.php\",\"line\":39,\"function\":\"handleRequest\",\"class\":\"OC\",\"type\":\"::\"}],\"File\":\"/var/www/nextcloud/lib/private/TextToImage/Manager.php\",\"Line\":231},\"message\":\"Call to a member function getExpectedRuntime() on null in file '/var/www/nextcloud/lib/private/TextToImage/Manager.php' line 231\",\"exception\":{},\"CustomMessage\":\"Call to a member function getExpectedRuntime() on null in file '/var/www/nextcloud/lib/private/TextToImage/Manager.php' line 231\"}",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "version": "28.0.1.0"
}
```

### PR DESCRIPTION
**Pull request by marcelklehr**
* Resolves: https://github.com/nextcloud/text2image_helper/issues/21

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by come-nc — COMMENTED** [review]
Why? Is there code relying on this? If so, please update the OCP return type accordingly in phpdoc to indicate `array<int,IProvider>` or `list<IProvider>` depending on what is needed.

[EDIT] Reading the referenced issue, you would need `list<IProvider>` to be sure that you can use key `0` without issue.

**Comment by MB-Finski** [comment]
> Why? Is there code relying on this? If so, please update the OCP return type accordingly in phpdoc to indicate `array<int,IProvider>` or `list<IProvider>` depending on what is needed.

The runOrScheduleTask assumes an indexed array with [0] being defined. This causes issues if a user has not specified the default provider and getPreferredProviders returns the unfiltered results of the getProviders() as it currently does. This seemed like the easiest place to fix this.

**Comment by MB-Finski** [comment]
@come-nc I hope I got it right. (I'm essentially a php newbie).

**Comment by marcelklehr** [comment]
Can't we just keep `IProvider[]`?

**Comment by come-nc** [comment]
> Can't we just keep `IProvider[]`?

You can but then calling code cannot assume that index 0 will be populated as the keys can be anything.

**Comment by marcelklehr** [comment]
Ah, I kind of assumed string[] to be a list<string>, but it's a array<array-key, string> :see_no_evil:

**Comment by come-nc** [comment]
> Ah, I kind of assumed string[] to be a list, but it's a array<array-key, string> 🙈

Yeah, `list` forces indexes to be contiguous so you cannot unset in the middle without reindexing, it’s restrictive to it’s not the default type.

File: lib/private/TextToImage/Manager.php [hunk 1]
```diff
@@ -83,7 +83,7 @@ public function getProviders(): array {
 		foreach ($context->getTextToImageProviders() as $providerServiceRegistration) {
 			$class = $providerServiceRegistration->getService();
 			try {
-				$this->providers[$class] = $this->serverContainer->get($class);
+				$this->providers[] = $this->serverContainer->get($class);
```
Comment (nickvergessen) [inline lib/private/TextToImage/Manager.php:86]:
```
lib/private/TextToImage/Manager.php:94:10: LessSpecificReturnStatement: The type 'list<mixed>' is more general than the declared return type 'list<OCP\TextToImage\IProvider>' for OC\TextToImage\Manager::getProviders (see https://psalm.dev/129)
```
Need to adjust the docs on the definition of the member as well

**Review by come-nc — APPROVED** [review]

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**